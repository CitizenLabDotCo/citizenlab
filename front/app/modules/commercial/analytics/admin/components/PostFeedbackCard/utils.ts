import { CornerRadius } from 'components/admin/Graphs/typings';

export const getCornerRadius =
  (stackLength: number, cornerRadius: number) =>
  ({ stackIndex }: { stackIndex: number }): CornerRadius => {
    const r = cornerRadius;
    if (stackIndex === 0) return [r, 0, 0, r];
    if (stackIndex === stackLength - 1) return [0, r, r, 0];

    return 0;
  };
