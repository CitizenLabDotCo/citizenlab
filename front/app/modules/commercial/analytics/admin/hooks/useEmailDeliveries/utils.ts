import { TimeSeriesResponse } from './typings';
import { IResolution } from 'components/admin/ResolutionControl';

export const deduceResolution = (
  timeSeriesResponse: TimeSeriesResponse
): IResolution | null => {
  if (timeSeriesResponse.length === 0) return null;
  const firstRow = timeSeriesResponse[0];

  if ('dimension_date_last_action.month' in firstRow) {
    return 'month';
  }

  if ('dimension_date_last_action.week' in firstRow) {
    return 'week';
  }

  return 'day';
};
