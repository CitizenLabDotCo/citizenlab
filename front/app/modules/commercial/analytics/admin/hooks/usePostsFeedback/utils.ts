import { Response, EmptyResponse } from '.';

export const isEmptyResponse = (
  response: Response | EmptyResponse
): response is EmptyResponse => {
  const [feedbackRows, statusRows] = response.data;
  const feedbackRow = feedbackRows[0];

  return (
    Object.values(feedbackRow).every((value) => value === null) ||
    statusRows.length === 0
  );
};
