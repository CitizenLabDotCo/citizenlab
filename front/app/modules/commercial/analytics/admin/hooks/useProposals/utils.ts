import { WrappedComponentProps } from 'react-intl';
import messages from '../../components/ProposalsCard/messages';
import { IResolution } from '../../../../../../components/admin/ResolutionControl';
import moment from 'moment';

export interface Labels {
  cardTitle: string;
  total: string;
  successful: string;
  successfulToolTip: string;
  period: string;
}

export const getLabels = (
  formatMessage: WrappedComponentProps['intl']['formatMessage'],
  resolution
): Labels => {
  let period = messages.last30Days;
  if (resolution === 'week') period = messages.last7Days;
  if (resolution === 'day') period = messages.yesterday;

  return {
    cardTitle: formatMessage(messages.proposals),
    total: formatMessage(messages.totalProposals),
    successful: formatMessage(messages.successfulProposals),
    successfulToolTip: formatMessage(messages.successfulProposalsToolTip),
    period: formatMessage(period),
  };
};

export const getLastPeriodMoment = (resolution: IResolution) => {
  let days = 30;
  if (resolution === 'week') days = 7;
  if (resolution === 'day') days = 1;
  return moment().subtract({ days });
};
