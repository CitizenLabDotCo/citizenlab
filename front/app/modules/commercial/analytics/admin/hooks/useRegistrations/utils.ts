import { TimeSeriesResponseRow } from './typings';
import { IResolution } from 'components/admin/ResolutionControl';

export const deduceResolution = (
  timeSeriesResponse: TimeSeriesResponseRow[]
): IResolution | null => {
  if (timeSeriesResponse.length === 0) return null;
  const firstRow = timeSeriesResponse[0];

  if ('dimension_date_registration.month' in firstRow) {
    return 'month';
  }

  if ('dimension_date_registration.week' in firstRow) {
    return 'week';
  }

  return 'day';
};
