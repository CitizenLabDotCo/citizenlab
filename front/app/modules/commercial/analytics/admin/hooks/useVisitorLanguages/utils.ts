// i18n
import messages from './messages';
import cardMessages from '../../components/VisitorsLanguageCard/messages';

// typings
import { WrappedComponentProps } from 'react-intl';

export interface Translations {
  language: string;
  count: string;
  title: string;
}

export const getTranslations = (
  formatMessage: WrappedComponentProps['intl']['formatMessage']
): Translations => ({
  language: formatMessage(messages.language),
  count: formatMessage(messages.count),
  title: formatMessage(cardMessages.title),
});
