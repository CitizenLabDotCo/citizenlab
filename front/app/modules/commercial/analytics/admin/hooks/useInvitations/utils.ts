import { WrappedComponentProps } from 'react-intl';
import messages from '../../components/InvitationsCard/messages';
import { getTimePeriodTranslationByResolution } from '../../utils/resolution';

export interface InvitationsLabels {
  cardTitle: string;
  fileName: string;
  total: string;
  pending: string;
  accepted: string;
  period: string;
}

export const getInvitationsLabels = (
  formatMessage: WrappedComponentProps['intl']['formatMessage'],
  resolution
): InvitationsLabels => {
  return {
    cardTitle: formatMessage(messages.invitations),
    fileName: formatMessage(messages.invitations)
      .toLowerCase()
      .replace(' ', '_'),
    total: formatMessage(messages.totalInvites),
    pending: formatMessage(messages.pending),
    accepted: formatMessage(messages.accepted),
    period: getTimePeriodTranslationByResolution(formatMessage, resolution),
  };
};
