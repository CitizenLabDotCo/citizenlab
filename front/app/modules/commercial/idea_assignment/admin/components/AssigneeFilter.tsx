import React, { SyntheticEvent } from 'react';

// import { Select } from '@citizenlab/cl2-component-library';
import { Dropdown } from 'semantic-ui-react';

import useAuthUser from 'api/me/useAuthUser';
import { IUsers } from 'api/users/types';
import useUsers from 'api/users/useUsers';

import { ManagerType } from 'components/admin/PostManager';
import postManagerMessages from 'components/admin/PostManager/messages';

import { trackEventByName } from 'utils/analytics';
import { useIntl } from 'utils/cl-intl';
import { isNilOrError } from 'utils/helperUtils';
import { getFullName } from 'utils/textUtils';

import messages from './messages';
import tracks from './tracks';

interface Props {
  handleAssigneeFilterChange: (value: string | undefined) => void;
  projectId?: string | null;
  assignee?: string | null;
  type: ManagerType;
  className?: string;
}

type TAssigneeOption = {
  value: string;
  text: string;
  id?: string;
  className?: string;
};

const AssigneeFilter = ({
  assignee,
  className,
  handleAssigneeFilterChange,
  type,
  projectId,
}: Props) => {
  const { formatMessage } = useIntl();
  const { data: authUser } = useAuthUser();
  const { data: prospectAssignees } = useUsers({
    pageSize: 250,
    ...((type === 'ProjectIdeas' || type === 'ProjectProposals') && projectId
      ? {
          can_moderate_project: projectId,
        }
      : type === 'AllIdeas'
      ? {
          can_moderate: true,
        }
      : { can_admin: true }),
  });

  if (!authUser || !prospectAssignees) {
    return null;
  }

  const getAssigneeOptions = (prospectAssignees: IUsers) => {
    // Order of assignee filter options:
    // All ideas > Assigned to me > Unassigned > Assigned to X (other admins/mods)
    const assigneeOptions: TAssigneeOption[] = [
      {
        value: 'all',
        text: formatMessage(messages.anyAssignment),
        id: 'e2e-assignee-filter-all-posts',
      },
      {
        value: authUser.data.id,
        text: formatMessage(postManagerMessages.assignedToMe),
        id: 'e2e-assignee-filter-assigned-to-user',
      },
      {
        value: 'unassigned',
        text: formatMessage(postManagerMessages.noOne),
        id: 'e2e-assignee-filter-unassigned',
      },
    ];

    const dynamicOptions = !isNilOrError(prospectAssignees)
      ? prospectAssignees.data
          .filter((assignee) => assignee.id !== authUser.data.id)
          .map((assignee) => ({
            value: assignee.id,
            text: formatMessage(postManagerMessages.assignedTo, {
              assigneeName: getFullName(assignee),
            }),
            className: 'e2e-assignee-filter-other-user',
          }))
      : [];

    return [...assigneeOptions, ...dynamicOptions];
  };

  const onAssigneeChange = (
    _event: SyntheticEvent,
    assigneeOption: TAssigneeOption
  ) => {
    const realFiterParam =
      assigneeOption.value === 'all' ? undefined : assigneeOption.value;

    trackEventByName(tracks.assigneeFilterUsed, {
      assignee: realFiterParam,
      adminAtWork: authUser.data.id,
    });

    handleAssigneeFilterChange(realFiterParam);
  };

  return (
    <Dropdown
      className={`${className} intercom-admin-asignee-filter`}
      id="e2e-select-assignee-filter"
      data-testid="assignee-filter-dropdown"
      options={getAssigneeOptions(prospectAssignees)}
      onChange={onAssigneeChange}
      value={assignee || 'all'}
      search
      selection
    />
  );
};

export default AssigneeFilter;
