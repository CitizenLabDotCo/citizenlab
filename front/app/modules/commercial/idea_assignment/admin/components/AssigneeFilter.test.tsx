// @ts-nocheck
import React from 'react';
import { shallow } from 'enzyme';

import { makeUser } from 'services/__mocks__/users';
import { intl } from 'utils/cl-intl';

// mocking dependencies
jest.mock('resources/GetUsers', () => 'GetUsers');
jest.mock('resources/GetAuthUser', () => 'GetAuthUser');
jest.mock('utils/cl-intl');
jest.mock('modules', () => ({ streamsToReset: [] }));

import { AssigneeFilter } from './AssigneeFilter';

describe('<AssigneeFilter />', () => {
  let handleAssigneeFilterChange: jest.Mock;
  beforeEach(() => {
    handleAssigneeFilterChange = jest.fn();
  });

  it('processes options correctly', () => {
    const authUser = makeUser({ roles: [{ type: 'admin' }] }, 'me').data;
    const prospectAssignees = ['admin1', 'admin2', 'admin3'].map(
      (name) =>
        makeUser(
          { slug: name, first_name: name, roles: [{ type: 'admin' }] },
          name
        ).data
    );
    prospectAssignees.push(authUser);

    const wrapper = shallow(
      <AssigneeFilter
        intl={intl}
        authUser={authUser}
        handleAssigneeFilterChange={handleAssigneeFilterChange}
        prospectAssignees={{ usersList: prospectAssignees }}
        assignee="me"
        projectId={undefined}
      />
    );
    expect(wrapper.find('Dropdown').prop('options')).toMatchSnapshot();
  });

  it('passes the user ID to handleAssigneeOnChange', () => {
    const authUser = makeUser({ roles: [{ type: 'admin' }] }, 'me').data;
    const prospectAssignees = ['admin1', 'admin2', 'admin3'].map(
      (name) =>
        makeUser(
          { slug: name, first_name: name, roles: [{ type: 'admin' }] },
          name
        ).data
    );
    prospectAssignees.push(authUser);

    const wrapper = shallow(
      <AssigneeFilter
        intl={intl}
        authUser={authUser}
        handleAssigneeFilterChange={handleAssigneeFilterChange}
        prospectAssignees={{ usersList: prospectAssignees }}
        assignee="me"
        projectId={undefined}
      />
    );
    const options = wrapper.find('Dropdown').prop('options');
    const pickedOption = options.find((option) => option.value === 'admin1');
    wrapper.instance().onAssigneeChange({}, pickedOption);
    expect(handleAssigneeFilterChange).toHaveBeenCalledTimes(1);
    expect(handleAssigneeFilterChange).toHaveBeenCalledWith('admin1');
  });

  it('passes down unassigned if you select Unassigned', () => {
    const authUser = makeUser({ roles: [{ type: 'admin' }] }, 'me').data;
    const prospectAssignees = ['admin1', 'admin2', 'admin3'].map(
      (name) =>
        makeUser(
          { slug: name, first_name: name, roles: [{ type: 'admin' }] },
          name
        ).data
    );
    prospectAssignees.push(authUser);

    const wrapper = shallow(
      <AssigneeFilter
        intl={intl}
        authUser={authUser}
        handleAssigneeFilterChange={handleAssigneeFilterChange}
        prospectAssignees={{ usersList: prospectAssignees }}
        assignee="me"
        projectId={undefined}
      />
    );
    const options = wrapper.find('Dropdown').prop('options');
    const unassigned = options.find((option) => option.value === 'unassigned');
    wrapper.instance().onAssigneeChange({}, unassigned);
    expect(handleAssigneeFilterChange).toHaveBeenCalledTimes(1);
    expect(handleAssigneeFilterChange).toHaveBeenCalledWith('unassigned');
  });
});
