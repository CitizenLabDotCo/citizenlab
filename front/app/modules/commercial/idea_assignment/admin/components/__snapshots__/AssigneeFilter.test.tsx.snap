// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssigneeFilter /> processes options correctly 1`] = `
Array [
  Object {
    "id": "e2e-assignee-filter-all-posts",
    "text": "Any administrator",
    "value": "all",
  },
  Object {
    "id": "e2e-assignee-filter-assigned-to-user",
    "text": "Assigned to me",
    "value": "me",
  },
  Object {
    "id": "e2e-assignee-filter-unassigned",
    "text": "Unassigned",
    "value": "unassigned",
  },
  Object {
    "className": "e2e-assignee-filter-other-user",
    "text": "Assigned to admin1 Citizenlab",
    "value": "admin1",
  },
  Object {
    "className": "e2e-assignee-filter-other-user",
    "text": "Assigned to admin2 Citizenlab",
    "value": "admin2",
  },
  Object {
    "className": "e2e-assignee-filter-other-user",
    "text": "Assigned to admin3 Citizenlab",
    "value": "admin3",
  },
]
`;
